source/core/Events.cpp

```diff
@@ -953,20 +1057,24 @@ VIREO_FUNCTION_SIGNATUREV(WaitForEventsAndDispatch, WaitForEventsParamBlock)
         InstructionCore *esBranchTarget = arguments[inputTuple].branchTarget;
         if (EventMatch(eventData, eventSpecRef[eventSpecIndex], dynIndex)) {
             // gPlatform.IO.Printf("Event match %d -> %x\n", eventSpecRef[eventSpecIndex].eventType, esBranchTarget);
-            const Int32 commonDataSize = sizeof(EventCommonData);
+            Int32 commonDataSize = sizeof(EventCommonData);
             Int32 dataNodeSize = esEventDataNodeType->TopAQSize();
             // TODO(spathiwa) Compute eventIndex when we support multiple events sharing event cases
             UInt32 eventIndex = 0;
             if (dataNodeSize >= commonDataSize) {
                 memcpy(esEventDataNode, &eventData, commonDataSize);
                 *EventIndexFieldPtr(esEventDataNode) = eventIndex;
-
-                if (eventData.eventDataType->TopAQSize() == dataNodeSize - commonDataSize) {
-                    // TODO(spathiwa) Should we also verify type of event data node matches event at run-time?
-                    // When we support multiple events with a shared event case, this needs to be smarter
-                    // and copy only common data, using a custom emit proc (which could verify types) and
-                    // generate custom snippets for each event data node cluster.
-                    eventData.eventDataType->CopyData(eventData.pEventData, (AQBlock1*)esEventDataNode + commonDataSize);
+                if (eventData.eventDataType) {
+                    Int32 alignment = eventData.eventDataType->AQAlignment();
+                    if (commonDataSize % alignment != 0)
+                        commonDataSize = (commonDataSize + alignment-1) & ~(alignment-1);  // round up to alignment boundary
+                    if (eventData.eventDataType->TopAQSize() == dataNodeSize - commonDataSize) {
+                        // TODO(spathiwa) Should we also verify type of event data node matches event at run-time?
```

Comment:
Is this something we need handle now? Or is this already handled and we can remove the TODO?

Reply:
It still needs to be handled.  Actually, implementing the common data copying will also necessarily have to do this check.  I think I may go ahead and implement it the straight-forward way instead of using a custom emit proc and then we can consider refactoring that way only if performance turns out to be an issue.